drivers/gnss/quectel_lcx6g.c

```diff
@@ -0,0 +1,754 @@
+	struct modem_chat_match dynamic_match;
+	struct modem_chat_script_chat dynamic_script_chat;
+	struct modem_chat_script dynamic_script;
+
+	/* Allocation for responses from GNSS modem */
+	union {
+		uint16_t fix_rate_response;
+		gnss_systems_t enabled_systems_response;
+		enum gnss_navigation_mode navigation_mode_response;
+	};
+
+	struct gnss_data data;
+	struct k_spinlock lock;
+};
+
+#define MODEM_CHAT_SCRIPT_NO_ABORT_DEFINE(_sym, _script_chats, _callback, _timeout)             \
+	static struct modem_chat_script _sym = {                                                \
+		.name = #_sym,                                                                  \
+		.script_chats = _script_chats,                                                  \
+		.script_chats_size = ARRAY_SIZE(_script_chats),                                 \
+		.abort_matches = NULL,                                                          \
+		.abort_matches_size = 0,                                                        \
+		.callback = _callback,                                                          \
+		.timeout = _timeout,                                                            \
+	}
+
+#ifdef CONFIG_PM_DEVICE
+MODEM_CHAT_MATCH_DEFINE(pair003_success_match, "$PAIR001,003,0*38", "", NULL);
+MODEM_CHAT_SCRIPT_CMDS_DEFINE(
+	suspend_script_cmds,
+	MODEM_CHAT_SCRIPT_CMD_RESP("$PAIR003*39", pair003_success_match)
+);
+
+MODEM_CHAT_SCRIPT_NO_ABORT_DEFINE(suspend_script, suspend_script_cmds,
+				  NULL, QUECTEL_LCX6G_SCRIPT_TIMEOUT_S);
+#endif /* CONFIG_PM_DEVICE */
+
+MODEM_CHAT_MATCH_DEFINE(any_match, "", "", NULL);
+MODEM_CHAT_MATCH_DEFINE(pair062_ack_match, "$PAIR001,062,0*3F", "", NULL);
+MODEM_CHAT_SCRIPT_CMDS_DEFINE(
+	resume_script_cmds,
+	MODEM_CHAT_SCRIPT_CMD_RESP("$PAIR002*38", any_match),
+	MODEM_CHAT_SCRIPT_CMD_RESP("$PAIR062,0,1*3F", pair062_ack_match),
+	MODEM_CHAT_SCRIPT_CMD_RESP("$PAIR062,1,0*3F", pair062_ack_match),
+	MODEM_CHAT_SCRIPT_CMD_RESP("$PAIR062,2,0*3C", pair062_ack_match),
+	MODEM_CHAT_SCRIPT_CMD_RESP("$PAIR062,3,5*38", pair062_ack_match),
+	MODEM_CHAT_SCRIPT_CMD_RESP("$PAIR062,4,1*3B", pair062_ack_match),
+	MODEM_CHAT_SCRIPT_CMD_RESP("$PAIR062,5,0*3B", pair062_ack_match),
+);
+
+MODEM_CHAT_SCRIPT_NO_ABORT_DEFINE(resume_script, resume_script_cmds,
+				  NULL, QUECTEL_LCX6G_SCRIPT_TIMEOUT_S);
+
+static void quectel_lcx6g_gga_cb(struct modem_chat *chat, char **argv, uint16_t argc,
+				 void *user_data)
+{
+	struct quectel_lcx6g_data *data = (struct quectel_lcx6g_data *)user_data;
+
+	if (gnss_nmea0183_parse_gga((const char **)argv, argc, &data->data) < 0) {
+		return;
+	}
+}
+
+static void quectel_lcx6g_rmc_cb(struct modem_chat *chat, char **argv, uint16_t argc,
+				 void *user_data)
+{
+	struct quectel_lcx6g_data *data = (struct quectel_lcx6g_data *)user_data;
+	int ret;
+
+	ret = gnss_nmea0183_parse_rmc((const char **)argv, argc, &data->data);
+	if (ret < 0) {
+		return;
+	}
+
+	gnss_publish_data(data->dev, &data->data);
+}
```

Comment:
hey I realize now, there's nothing really device specific in these apart from the definition of the data struct, would it make sense to have some generic versions of these defined at library level that can be reused by multiple NMEA based drivers if they don't need any device specific quirks? Would require some rule about `data` being first in the struct or something but it may save some duplicated code.

Reply:
I would say so, I will be adding a bit more logic to these to ensure a valid GGA and RMC message has been received before publishing the data (they can occur in different order on different modems for example), and some logic will need to be added to track the GSV messages as well. I will create a "base" for NMEA based GNSS modems :)

Reply:
good stuff, sounds reasonable